src/code/sys_flashrom.c

```diff
@@ -3,42 +3,242 @@
+    data = SystemArena_Malloc(size);
+    if (data == 0) {
+        ret = SysFlashrom_AttemptWrite(addr, pageNum, pageCount);
+    } else {
+        SysFlashrom_ReadData(data, pageNum, pageCount);
+        if (bcmp(data, addr, size) == 0) {
+            ret = 0;
+        } else {
+            //! @bug Will attempt to write regardless of the result of the function call.
+            if (func_80185BE4(data, addr, pageCount) != 0) {
+                ret = SysFlashrom_AttemptWrite(addr, pageNum, pageCount);
+            } else {
+                ret = SysFlashrom_AttemptWrite(addr, pageNum, pageCount);
+            }
+            if (ret == 0) {
+                SysFlashrom_ReadData(data, pageNum, pageCount);
+                if (bcmp(data, addr, size) == 0) {
+                    ret = 0;
+                } else {
+                    ret = -1;
+                }
+            }
+        }
+        SystemArena_Free(data);
+    }
+    return ret;
+}
+
+void SysFlashrom_ThreadEntry(void* arg) {
+    FlashromRequest* req = (FlashromRequest*)arg;
+    switch (req->requestType) {
+        case 1:
+            req->response = SysFlashrom_WriteData2(req->addr, req->pageNum, req->pageCount);
+            osSendMesg(&req->messageQueue, req->response, OS_MESG_BLOCK);
+            break;
+        case 2:
+            req->response = SysFlashrom_ReadData(req->addr, req->pageNum, req->pageCount);
+            osSendMesg(&req->messageQueue, req->response, OS_MESG_BLOCK);
+            break;
+    }
+}
+
+void SysFlashrom_CreateRequest(u8* addr, u32 pageNum, u32 pageCount) {
+    FlashromRequest* req = &sFlashromRequest;
+    if (SysFlashrom_FlashIsGood() != 0) {
+        req->requestType = 1;
+        req->addr = addr;
+        req->pageNum = pageNum;
+        req->pageCount = pageCount;
+        osCreateMesgQueue(&req->messageQueue, sSysFlashromMsgBuf, ARRAY_COUNT(sSysFlashromMsgBuf));
+        StackCheck_Init(&sSysFlashromStackInfo, sSysFlashromStack, STACK_TOP(sSysFlashromStack), 0,
+                        0x100, "sys_flashrom");
+        osCreateThread(&sSysFlashromThread, Z_THREAD_ID_FLASHROM, SysFlashrom_ThreadEntry, req,
+                       STACK_TOP(sSysFlashromStack), Z_PRIORITY_FLASHROM);
+        osStartThread(&sSysFlashromThread);
+    }
+}
+
+s32 SysFlashrom_IsQueueFull(void) {
+    OSMesgQueue* queue = &sFlashromRequest.messageQueue;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    return MQ_IS_FULL(queue); //queue->msgCount <= queue->validCount;
+}
+
+s32 SysFlashrom_DestroyThread(void) {
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    osRecvMesg(&sFlashromRequest.messageQueue, NULL, OS_MESG_BLOCK);
+    osDestroyThread(&sSysFlashromThread);
+    StackCheck_Cleanup(&sSysFlashromStackInfo);
+    return sFlashromRequest.response;
+}
+
+void func_80185F64(void* addr, u32 pageNum, u32 pageCount) {
+    SysFlashrom_CreateRequest(addr, pageNum, pageCount);
+    SysFlashrom_DestroyThread();
+}
```

Comment:
```suggestion
s32 SysFlashrom_WriteDataSync(void* addr, u32 pageNum, u32 pageCount) {
    SysFlashrom_CreateRequest(addr, pageNum, pageCount);
    return SysFlashrom_DestroyThread();
}
```
Since it issues a write request then immediately waits for the result, which blocks the calling thread while just running the first function does not block the calling thread.